snapshots/manager.go

```diff
@@ -50,33 +31,59 @@ type restoreDone struct {
 //  2. io.ReadCloser streams automatically propagate IO errors, and can pass arbitrary
 //     errors via io.Pipe.CloseWithError().
 type Manager struct {
-	store      *Store
-	multistore types.Snapshotter
 	extensions map[string]types.ExtensionSnapshotter
+	// store is the snapshot store where all completed snapshots are persisted.
+	store *Store
+	opts  types.SnapshotOptions
+	// multistore is the store from which snapshots are taken.
+	multistore types.Snapshotter
+	logger     log.Logger
+
+	mtx               sync.Mutex
+	operation         operation
+	chRestore         chan<- uint32
+	chRestoreDone     <-chan restoreDone
+	restoreSnapshot   *types.Snapshot
+	restoreChunkIndex uint32
+}
+
+// operation represents a Manager operation. Only one operation can be in progress at a time.
+type operation string
 
-	mtx                sync.Mutex
-	operation          operation
-	chRestore          chan<- io.ReadCloser
-	chRestoreDone      <-chan restoreDone
-	restoreChunkHashes [][]byte
-	restoreChunkIndex  uint32
+// restoreDone represents the result of a restore operation.
+type restoreDone struct {
+	complete bool  // if true, restore completed successfully (not prematurely)
+	err      error // if non-nil, restore errored
 }
 
+const (
+	opNone     operation = ""
+	opSnapshot operation = "snapshot"
+	opPrune    operation = "prune"
+	opRestore  operation = "restore"
+
+	chunkBufferSize   = 4
+	chunkIDBufferSize = 1024
+
+	// snapshotMaxItemSize limits the size of both KVStore entries and snapshot
+	// extension payloads during a state-sync restore.
+	// Unexported so copied in manager_test.go for testing
+	snapshotMaxItemSize = int(512e6)
+)
+
+var ErrOptsZeroSnapshotInterval = errors.New("snapshot-interval must not be 0")
```

Comment:
Nit: upstream has a typo, I'd prefer if we kept it.


```suggestion
var ErrOptsZeroSnapshotInterval = errors.New("snaphot-interval must not be 0")
```

Reply:
Quite right - I thought we had introduced that. Done.